states: 1: Update LifecycleOnce to have multiple possible states

willhug opened the pull request:
Summary: LifecycleOnce had a few concurrency gaps described in #634
This pr updates the LifecycleOnce interface to have a safe concurrency
story and has tests to validate that.  This involved adding a mutex
accross both Start and Stop and adding a 'state' atomic int to represent
States (Idle, Starting, Running, Stopping, Stopped, Errored)

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+	startOnce sync.Once
+	startErr  error
+	stopOnce  sync.Once
+	stopErr   error
```
I believe you only need a single atomic for the state. The atomic state is your lock. I am not sure about the utility of storing the errors?

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+	defer l.lock.Unlock()
+
+	if LifecycleState(l.state.Load()) != Idle {
+		return l.startErr
```
You cannot assume that the current error was about starting nor stopping. Technically, you cannot start because its not IDLE, so you can return an error describing that the precondition failed.

willhug replied:
Ah, so in one of the previous diffs we decided to make the Start/Stop functions idempotent, thus if you call them multiple times we won't run the underlying start/stop functions multiple times (which could have consequences depending on implementations).  This means that if the state is not idle, we have already run either start or stop, and we should return the error that happened on that first call.

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+		l.state.Store(int32(Errored))
+	}
+
+	return l.startErr
```
I am not sure what is the purpose of storing the startErr for later?

abhinav replied:
So that if you call Start again after it errors out, you get back the same error.

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
 func (l *LifecycleOnce) Start(f func() error) error {
-	return l.start.Do(f)
+	l.lock.Lock()
+	defer l.lock.Unlock()
```
no need for mutex here, the atomic state is your lock

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
 func (l *LifecycleOnce) Stop(f func() error) error {
-	return l.stop.Do(f)
+	l.lock.Lock()
+	defer l.lock.Unlock()
```
similar as above, no need for the mutex

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+	defer l.lock.Unlock()
+
+	// If the lifecycle is already stopped or errored, don't execute the stop func
+	if LifecycleState(l.state.Load()) == Stopped || LifecycleState(l.state.Load()) == Errored {
```
you are doing two load here, this is a race condition. Even if you load only once, you have a race condition. You need to instead do a compare and swap to STOPPING.

willhug replied:
There is no race condition because I'm in a mutex lock

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	if LifecycleState(l.state.Load()) != Idle {
```
you need to do a compare and swap with STARTING, else you have a race condition.

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+		f = func() error { return nil }
+	}
+
+	l.state.Store(int32(Starting))
```
race condition here, should be compare and swapped above.

willhug replied:
mutex, all good!

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+		l.state.Store(int32(Running))
+	} else {
+		l.state.Store(int32(Errored))
+	}
```
here you are good to Store directly, as "own the lock" by have won the race to compare and swap to STARTING.

willhug replied:
Mostly the reason why state is in an atomic is because though you won't be writing to it at the same time, it is possible to read from it using the `IsRunning` function (or another `State` function we might show)

abhinav replied:
It may be simpler to use an RWLock and acquire the read-lock in IsRunning

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+		f = func() error { return nil }
+	}
+
+	l.state.Store(int32(Stopping))
```
this is handled by the cmpswap above.

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+		l.state.Store(int32(Errored))
+	}
+
+	return l.stopErr
```
same question as for start, why storing the stopErr in the struct?

bombela commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,105 @@
+	state := LifecycleState(l.state.Load())
+	return state == Starting ||
+		state == Running ||
+		state == Stopping
```
And when every move to *ING states is cmpswap, this will become consistent.

willhug commented:
I addressed some of the comments, but I don't think I'll have time to go through a review cycle because I'm going to be on vacation tomorrow.  I don't think any changes we make on the lifecycleOnce object should affect 1.0 since it's hidden in an internal package, and the old version is still functional.  This is only for the case where we start weaving start/stops of inbounds/outbounds/etc in weird ways.

@abhinav @bombela feel free to coordinate and take this from me if you wish.

abhinav commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,111 @@
+)
+
+// LifecycleState represents `states` that a lifecycle object can be in.
+type LifecycleState int
```
Not sure this and others need to be public

abhinav commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,111 @@
+	}
+
+	if f == nil {
+		f = func() error { return nil }
```
nit: create a module-level `nopFunc` which you use here and in start

abhinav commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,111 @@
+
+	// If the lifecycle is already stopped or errored, don't execute the stop func
+	if LifecycleState(l.state.Load()) == Stopped || LifecycleState(l.state.Load()) == Errored {
+		return l.stopErr
```
if the state is errored because we called start, we need to return that error here instead

maybe we should use a single err attribute

abhinav commented on internal/sync/lifecycle.go:
```diff
@@ -20,24 +20,111 @@
-	return l.start.Done() && !l.stop.Done()
+	state := LifecycleState(l.state.Load())
+	return state == Starting ||
+		state == Running ||
```
isn't only Running == IsRunning?

bombela commented:
To be clear all my comments where assuming you get rid of the mutex, because why bother with a mutex if you already have the atomic to make progression. This is also the only reason I proposed the intermediate *ING states in the first place, in order to have a safe & atomic progression. With a mutex you don't need the intermediate state, unless you want to expose it for debugging/introspection of course.

Because its a slow path, the mutex is not a performance concern, but remember, this adds one mutex to every lifecycle object. Every atomic Load/Store also takes few cycle because it has to sync the cache across CPU cores. Sure this wastes only few 100 bytes of RAM forever and few cycles of CPU when a process start. But less is better no? ;)

The current code has a IsRunning() which is not consistent and you can call Stop() before Start(), and then Stop() will never do anything anymore. Now its probably not a big deal, but just so you know, I encountered the bug as I was testing introspection. To be fair it was a twisted test anyway. (I wanted to double check that my Dispatcher was properly registered/unregistered from a global list even if somebody mistakenly call Stop before Start). So probably not something urgent.

In conclusion, its up to you if you want to get rid of the mutex or not.

abhinav commented:
@bombela added a TODO to eliminate that later. This gives the desired behavior for now and it's completely internal.
